src/QsCompiler/SyntaxProcessor/SymbolOccurrence.fs

```diff
@@ -0,0 +1,194 @@
+        | QubitRegisterAllocation e -> inExpression e
+        | QubitTupleAllocation xs -> Seq.collect inInitializer xs |> Seq.toList
+        | SingleQubitAllocation
+        | InvalidInitializer -> []
+
+    let inGenerator generator =
+        match generator.Generator with
+        | Intrinsic
+        | AutoGenerated
+        | FunctorGenerationDirective _ -> []
+        | UserDefinedImplementation s -> symbolDeclarations s
+
+    let rec inNamedTuple =
+        function
+        | QsTuple xs -> Seq.collect inNamedTuple xs |> Seq.toList
+        | QsTupleItem (s, t) -> symbolDeclarations s @ inType t
+
+    let inCallable (callable: CallableDeclaration) =
+        let typeParams = Seq.map Declaration callable.Signature.TypeParameters |> Seq.toList
+        let args = inNamedTuple callable.Signature.Argument
+        let returnType = inType callable.Signature.ReturnType
+        Declaration callable.Name :: typeParams @ args @ returnType
+
+    [<CompiledName "InFragment">]
+    let inFragment fragment =
+        match fragment with
+        | NamespaceDeclaration s -> [ Declaration s ]
+        | OpenDirective (_, s) -> QsNullable<_>.Choose id [ s ] |> Seq.toList |> List.map Declaration
+        | DeclarationAttribute (s, e) -> UsedVariable s :: inExpression e
+        | OperationDeclaration c
+        | FunctionDeclaration c -> inCallable c
+        | BodyDeclaration g
+        | AdjointDeclaration g
+        | ControlledDeclaration g
+        | ControlledAdjointDeclaration g -> inGenerator g
+        | ImmutableBinding (s, e)
+        | MutableBinding (s, e)
+        | ForLoopIntro (s, e) -> symbolDeclarations s @ inExpression e
+        | UsingBlockIntro (s, i)
+        | BorrowingBlockIntro (s, i) -> symbolDeclarations s @ inInitializer i
+        | TypeDefinition t -> Declaration t.Name :: inNamedTuple t.UnderlyingType
+        | ExpressionStatement e
+        | ReturnStatement e
+        | FailStatement e
+        | IfClause e
+        | ElifClause e
+        | WhileLoopIntro e
+        | UntilSuccess (e, _) -> inExpression e
+        | ValueUpdate (e1, e2) -> inExpression e1 @ inExpression e2
+        | ElseClause
+        | RepeatIntro
+        | WithinBlockIntro
+        | ApplyBlockIntro
+        | InvalidFragment -> []
+
+// C# interoperability.
+type SymbolOccurrence with
+    member occurrence.Match
+        (
+            declaration: Func<QsSymbol, 'a>,
+            usedType: Func<QsType, 'a>,
+            usedVariable: Func<QsSymbol, 'a>,
+            usedLiteral: Func<QsExpression, 'a>
+        ) =
+        match occurrence with
+        | Declaration s -> declaration.Invoke s
+        | UsedType t -> usedType.Invoke t
+        | UsedVariable s -> usedVariable.Invoke s
+        | UsedLiteral e -> usedLiteral.Invoke e
+
+    [<MaybeNull>]
+    member occurrence.AsDeclaration =
+        match occurrence with
+        | Declaration s -> s
+        | _ -> Unchecked.defaultof<QsSymbol>
```

Comment:
I would suggest to stick with returning a QsNullable instead. See also here, and the tldr: https://stackoverflow.com/questions/13520720/unchecked-defaultof-on-a-list.

Reply:
This is part of the C# interop API for SymbolOccurrence, so it should be as idiomatic to use from C# as possible. That's why I used a nullable type. The IL for the default case here is:

```
    IL_001f: ldnull
    IL_0020: ret
```

That should be equivalent to a C# method with a `SymbolOccurrence?` return type returning null, which is what I intended.

I don't think this will cause issues like that StackOverflow question, because in that question they are assigning `null` to a field of an F# list, which has a special interpretation. Here, `null` is only used as a return type for a method that is called from C#. One snag is that it does assume that `SymbolOccurrence` is a reference type; I don't think this will work if it has the `[<Struct>]` attribute.

Reply:
There are a couple Unchecked.defaultof in this file, and I am just a bit worried that ultimately, we might change one of the types or one of these methods will be used from F#, and this will lead to odd issues that go unnoticed. It just feels a bit unsafe, and I think the "use with caution" disclaimer (Unchecked) overall has good reason. I understand the desire for keeping it idiomatic, but this feels like one check too little especially for a public API. Plus, it doesn't align well with the rest of the code base, and I would have concerns using it broadly across the code base.

Reply:
It looks like the [F# component design guidelines recommend the TryGet pattern](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/component-design-guidelines#use-the-trygetvalue-pattern-instead-of-returning-f-option-values-and-prefer-method-overloading-to-taking-f-option-values-as-arguments) for C# interop, as the replacement for returning an option type.

Reply:
I'd prefer to go with that then. That seems reasonable to me - what do you think?

Reply:
I updated this to use the TryGet pattern.